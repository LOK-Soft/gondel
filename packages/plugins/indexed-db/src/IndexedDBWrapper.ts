import { IndexedDB, Transaction } from "./NativeIndexedDB";
import { VERSION } from "./Configuration";

export interface IIndexedDBEntry {
  id: string;
}

class IndexedDBWrapperError extends Error {
  constructor(method: string, message?: string) {
    super(`Transaction:${method}: ${message || "unknown error occured"}`);
  }
}

const throwOnInvalidConnection = (
  database: IDBDatabase | undefined,
  method: string,
  reject: (err: IndexedDBWrapperError) => void
) => {
  if (!database) {
    reject(
      new IndexedDBWrapperError(
        method,
        `connection is not established, probably connect() call missing`
      )
    );
  }
};

export class IndexedDBWrapper {
  private database: IDBDatabase;

  constructor(private name: string, private version: number = VERSION) {}

  public async connect(): Promise<any> {
    return this.establishInternalConnection();
  }

  public get<T extends Object>(idOrEntry: string | IIndexedDBEntry): Promise<T & IIndexedDBEntry> {
    return new Promise((resolve, reject) => {
      console.log("get", this.database);
      throwOnInvalidConnection(this.database, "get", reject);

      const transaction: IDBRequest = this.database
        .transaction([this.name])
        .objectStore(this.name)
        .get(typeof idOrEntry === "string" ? idOrEntry : idOrEntry.id);

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "get",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };

      transaction.onsuccess = () => {
        resolve(transaction.result as T & IIndexedDBEntry);
      };
    });
  }

  public set<T extends IIndexedDBEntry>(data: T): Promise<void> {
    return new Promise((resolve, reject) => {
      console.log("set", this.database);
      throwOnInvalidConnection(this.database, "set", reject);

      const transaction = this.database
        .transaction([this.name], "readwrite")
        .objectStore(this.name)
        .add(data);

      transaction.onsuccess = () => {
        resolve();
      };

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "set",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };
    });
  }

  public delete(idOrEntry: string | IIndexedDBEntry): Promise<void> {
    return new Promise((resolve, reject) => {
      throwOnInvalidConnection(this.database, "delete", reject);
      const transaction = this.database
        .transaction([this.name], "readwrite")
        .objectStore(this.name)
        .delete(typeof idOrEntry === "string" ? idOrEntry : idOrEntry.id);

      transaction.onsuccess = () => {
        resolve();
      };

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "delete",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };
    });
  }

  public __dangerouslyAccessInternalDatabase(): IDBDatabase {
    return this.database;
  }

  private establishInternalConnection(): Promise<IDBDatabase> {
    return new Promise((resolve, reject) => {
      const self = this;
      const transaction = IndexedDB.open(this.name, this.version);

      transaction.onerror = () => {
        reject(
          new IndexedDBWrapperError(
            "connect",
            transaction.error ? transaction.error.message : undefined
          )
        );
      };

      transaction.onsuccess = (event: Event) => {
        this.database = transaction.result;

        resolve(this.database);
      };

      transaction.onupgradeneeded = function(event: IDBVersionChangeEvent) {
        if (event.newVersion && event.oldVersion < event.newVersion) {
          ((event.target as any).result as IDBDatabase).createObjectStore(self.name, {
            keyPath: "id"
          });
        }
      };
    });
  }
}
