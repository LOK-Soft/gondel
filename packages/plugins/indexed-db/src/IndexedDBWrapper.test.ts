import { IndexedDBWrapper } from "./IndexedDBWrapper";
const { reset } = require("shelving-mock-indexeddb");

// Reset the IndexedDB mock before/after tests.
// This will clear all object stores, indexes, and data.
beforeEach(() => reset());
afterEach(() => reset());

// The IndexedDB mock uses setTimeout() to simulate the asyncronous API.
// Add fake timers before/after tests to ensure the asyncronous responses are received by the test.
beforeEach(() => jest.useFakeTimers());
afterEach(() => jest.runAllTimers());

describe("@gondel/plugin-indexed-db", () => {
  describe("IndexedDBWrapper", () => {
    it("should create a connection", async done => {
      expect.assertions(0);
      const db = new IndexedDBWrapper("testCase1");
      await db.connect();
      expect(db).toMatchSnapshot();
      expect(db.__dangerouslyAccessInternalDatabase()).toMatchSnapshot();
      done();
    });

    it(
      "should set/get values from the DB",
      async done => {
        expect.assertions(3);
        const db = new IndexedDBWrapper("testCase2");
        await db.connect();
        expect(db).toMatchSnapshot();
        expect(db.__dangerouslyAccessInternalDatabase()).toMatchSnapshot();

        await db.set({
          id: "1",
          value: "Hello World"
        });
        expect(db).toMatchSnapshot();
        expect(db.__dangerouslyAccessInternalDatabase()).toMatchSnapshot();

        const result = await db.get<{ value: string }>("1");
        expect(result.value).toEqual("Hello World");
        expect(db).toMatchSnapshot();
        expect(db.__dangerouslyAccessInternalDatabase()).toMatchSnapshot();
        done();
      },
      15000
    );
  });
});
