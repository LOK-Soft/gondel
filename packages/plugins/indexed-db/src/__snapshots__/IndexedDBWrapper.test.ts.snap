// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@gondel/plugin-indexed-db IndexedDBWrapper should create a connection 1`] = `
IndexedDBWrapper {
  "database": IDBDatabase {
    "name": "testCase1",
    "objectStoreNames": Array [
      "testCase1",
    ],
    "version": 1,
  },
  "name": "testCase1",
  "version": 1,
}
`;

exports[`@gondel/plugin-indexed-db IndexedDBWrapper should create a connection 2`] = `
IDBDatabase {
  "name": "testCase1",
  "objectStoreNames": Array [
    "testCase1",
  ],
  "version": 1,
}
`;

exports[`@gondel/plugin-indexed-db IndexedDBWrapper should set/get values from the DB 1`] = `
IndexedDBWrapper {
  "database": IDBDatabase {
    "name": "testCase2",
    "objectStoreNames": Array [
      "testCase2",
    ],
    "version": 1,
  },
  "name": "testCase2",
  "version": 1,
}
`;

exports[`@gondel/plugin-indexed-db IndexedDBWrapper should set/get values from the DB 2`] = `
IDBDatabase {
  "name": "testCase2",
  "objectStoreNames": Array [
    "testCase2",
  ],
  "version": 1,
}
`;
